### PR TITLE
[TS migration] Migrate 'ReportActionItemFragment.js' and related components to TypeScript

### PR DESCRIPTION
**Pull request by ishpaul777**
### Details

TS migration for ReportActionItemFragment.js

### Fixed Issues

$ https://github.com/Expensify/App/issues/31977
PROPOSAL: https://github.com/Expensify/App/issues/31977#issuecomment-1860925429

### Tests

- [x] Verify that no errors appear in the JS console

1. Login to the app
2. Open any report
3. Verify that report actions are displayed correctly (as before)
4. Send different types of messages Attachements/ text messages/ mentions etc. Make sure everything works as expected (as before)

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/104348397/a3756ed0-8038-4b12-b833-19d824f1d4ed

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/104348397/ee1439b2-70c6-440a-9783-af6f2af6016e

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/104348397/49a51743-350d-435a-b876-5d0e71ec9baa

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/104348397/b896f40c-0ba8-4c7a-b2d9-922cb843ccde

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/104348397/6b51337e-ebce-44ac-a51b-3954f098e2ff

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/104348397/29815ede-7843-49ab-9ee6-a10f1d618994

</details>

### CODE REVIEW

**Comment by blazejkustra** [comment]
I'll leave comments tomorrow morning as I'm finishing for today, sorry!

File: src/pages/home/report/comment/TextCommentFragment.tsx [hunk 1]
```diff
@@ -72,10 +65,13 @@ function TextCommentFragment(props) {
         );
     }
 
+    const propsStyle = Array.isArray(props.style) ? props.style : [props.style];
```
Comment (blazejkustra) [inline src/pages/home/report/comment/TextCommentFragment.tsx:68]:
```suggestion
```

File: src/pages/home/report/comment/TextCommentFragment.tsx [hunk 2]
```diff
@@ -72,10 +65,13 @@ function TextCommentFragment(props) {
...
+
     const containsOnlyEmojis = EmojiUtils.containsOnlyEmojis(text);
+    const message = isEmpty(iouMessage) ? text : iouMessage;
 
     return (
-        <Text style={[containsOnlyEmojis ? styles.onlyEmojisText : undefined, styles.ltr, ...props.style]}>
+        <Text style={[containsOnlyEmojis ? styles.onlyEmojisText : undefined, styles.ltr, ...propsStyle]}>
```
Comment (blazejkustra) [inline src/pages/home/report/comment/TextCommentFragment.tsx:74]: RN styles handles falsy values, as well as nested objects:
```suggestion
        <Text style={[containsOnlyEmojis && styles.onlyEmojisText, styles.ltr, style]}>
```

File: src/pages/home/report/comment/TextCommentFragment.tsx [hunk 3]
```diff
@@ -84,12 +80,12 @@ function TextCommentFragment(props) {
                 style={[
                     containsOnlyEmojis ? styles.onlyEmojisText : undefined,
                     styles.ltr,
-                    ...props.style,
+                    ...propsStyle,
```
Comment (blazejkustra) [inline src/pages/home/report/comment/TextCommentFragment.tsx:83]: Same here

File: src/pages/home/report/comment/TextCommentFragment.tsx [hunk 4]
```diff
@@ -102,18 +98,16 @@ function TextCommentFragment(props) {
                     <Text
                         fontSize={variables.fontSizeSmall}
                         color={theme.textSupporting}
-                        style={[styles.editedLabelStyles, styleAsDeleted ? styles.offlineFeedback.deleted : undefined, ...props.style]}
+                        style={[styles.editedLabelStyles, styleAsDeleted ? styles.offlineFeedback.deleted : undefined, ...propsStyle]}
```
Comment (blazejkustra) [inline src/pages/home/report/comment/TextCommentFragment.tsx:101]: Same here

File: src/pages/home/report/ReportActionItemMessage.tsx [hunk 5]
```diff
@@ -76,8 +76,8 @@ function ReportActionItemMessage({action, displayAsGroup, reportID, style, isHid
                 isThreadParentMessage={ReportActionsUtils.isThreadParentMessage(action, reportID)}
                 attachmentInfo={action.attachmentInfo}
                 pendingAction={action.pendingAction}
-                source={(action.originalMessage as OriginalMessageAddComment['originalMessage'])?.source}
-                accountID={action.actorAccountID}
+                source={action.originalMessage as OriginalMessageSource}
+                accountID={action.actorAccountID ?? 0}
                 style={style}
```
Comment (blazejkustra) [inline src/pages/home/report/ReportActionItemMessage.tsx:81]: Change `ReportActionItemMessageProps` to:
```ts
    /** Additional styles to add after local styles. */
    style?: StyleProp<ViewStyle & TextStyle>;
```

File: src/pages/home/report/ReportActionItemSingle.tsx [hunk 6]
```diff
@@ -256,7 +256,7 @@ function ReportActionItemSingle({
                                 <ReportActionItemFragment
                                     // eslint-disable-next-line react/no-array-index-key
                                     key={`person-${action.reportActionID}-${index}`}
-                                    accountID={actorAccountID}
+                                    accountID={actorAccountID ?? 0}
                                     fragment={fragment}
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply src/pages/home/report/ReportActionItemSingle.tsx:242]: I think it's safe to change Person type to:
```ts
type Person = {
    type: string;
    style?: string;
    text: string;
};
```

File: src/pages/home/report/ReportActionItemSingle.tsx [hunk 7]
```diff
@@ -238,7 +238,7 @@ function ReportActionItemSingle({
...
                                     delegateAccountID={action.delegateAccountID}
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply src/pages/home/report/ReportActionItemSingle.tsx:243]: I tried to console.log(delegateAccountID), but with no luck. It was always undefined. Can you try to find if it's number or string in reality? If not try asking in open source channel

**Review by blazejkustra — CHANGES_REQUESTED** [review]

**Comment by blazejkustra** [comment]
@ishpaul777 Typecheck is failing

**Comment by blazejkustra** [comment]
> Also a kind request, link issue in PR description and change PR title so it reflects issue title: [TS migration] Migrate 'X.js' component to TypeScript

bump 😄

File: src/pages/home/report/comment/TextCommentFragment.tsx [hunk 8]
```diff
@@ -1,56 +1,48 @@
...
-};
-
-const defaultProps = {
-    iouMessage: undefined,
+    /** Text of an IOU report action */
+    iouMessage?: string;
 };
 
-function TextCommentFragment(props) {
+function TextCommentFragment({fragment, styleAsDeleted, source, style, displayAsGroup, iouMessage = ''}: TextCommentFragmentProps) {
```
Comment (blazejkustra) [inline src/pages/home/report/comment/TextCommentFragment.tsx:40]: Is this default type needed?
```suggestion
function TextCommentFragment({fragment, styleAsDeleted, source, style, displayAsGroup, iouMessage}: TextCommentFragmentProps) {
```

**Review by blazejkustra — APPROVED** [review]
One minor comment, other than that LGTM!

**Comment by blazejkustra** [comment]
FYI Lint is failing

**Comment by aldo-expensify** [comment]
Reassigned to @situchan as C+
Assigned @jasperhuangg as reviewer since he is the assigned engineer in the [issue](https://github.com/Expensify/App/issues/31977)

**Comment by aldo-expensify** [comment]
@ishpaul777 you got some conflicts now

File: src/pages/home/report/ReportActionItemMessage.tsx [hunk 9]
```diff
@@ -74,10 +74,9 @@ function ReportActionItemMessage({action, displayAsGroup, reportID, style, isHid
                 fragment={fragment}
                 iouMessage={iouMessage}
                 isThreadParentMessage={ReportActionsUtils.isThreadParentMessage(action, reportID)}
-                attachmentInfo={action.attachmentInfo}
                 pendingAction={action.pendingAction}
-                source={(action.originalMessage as OriginalMessageAddComment['originalMessage'])?.source}
-                accountID={action.actorAccountID}
+                source={action.originalMessage as OriginalMessageSource}
+                accountID={action.actorAccountID ?? 0}
```
[the remark this answers is not in the document]
Comment (jasperhuangg) [reply src/pages/home/report/ReportActionItemMessage.tsx:82]: Is this resolved? Seems so

File: src/pages/home/report/comment/TextCommentFragment.tsx [hunk 10]
```diff
@@ -66,54 +58,53 @@ function TextCommentFragment(props) {
...
                     styleAsDeleted ? styles.offlineFeedback.deleted : undefined,
-                    !DeviceCapabilities.canUseTouchScreen() || !props.isSmallScreenWidth ? styles.userSelectText : styles.userSelectNone,
+                    !DeviceCapabilities.canUseTouchScreen() || !isSmallScreenWidth ? styles.userSelectText : styles.userSelectNone,
                 ]}
             >
-                {convertToLTR(props.iouMessage || text)}
+                {convertToLTR(message)}
             </Text>
-            {Boolean(fragment.isEdited) && (
+            {!!fragment.isEdited && (
```
Comment (jasperhuangg) [inline src/pages/home/report/comment/TextCommentFragment.tsx:87]: Interesting–what else could this be besides a Boolean? I wonder why we decided to do this originally.

**Review by jasperhuangg — COMMENTED** [review]
Looks pretty good, let's fix those conflicts!

**Review by jasperhuangg — APPROVED** [review]

**Comment by situchan** [comment]
@ishpaul777 please merge main since 1k+ commits are behind

**Comment by blazejkustra** [comment]
Please sync this branch with the newest main. Context [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1706196639925099), thanks!

File: src/pages/home/report/ReportActionItemMessage.tsx [hunk 11]
```diff
@@ -77,10 +77,9 @@ function ReportActionItemMessage({action, displayAsGroup, reportID, style, isHid
                 fragment={fragment}
                 iouMessage={iouMessage}
                 isThreadParentMessage={ReportActionsUtils.isThreadParentMessage(action, reportID)}
-                attachmentInfo={action.attachmentInfo}
                 pendingAction={action.pendingAction}
-                source={(action.originalMessage as OriginalMessageAddComment['originalMessage'])?.source}
-                accountID={action.actorAccountID}
+                source={action.originalMessage as OriginalMessageSource}
```
Comment (situchan) [inline src/pages/home/report/ReportActionItemMessage.tsx:81]: sorry forgot to post comment. Why `.source` is removed here?

Comment (situchan) [reply src/pages/home/report/ReportActionItemMessage.tsx:81]: You might have confused with type. It's value.
Before: action.originalMessage?.source
After: action.originalMessage

**Pull request merged**